jax/lax/lax.py

```diff
@@ -735,7 +735,8 @@ def _get_min_identity(dtype):
     return onp.array(True, onp.bool_)
 
 def _reduce_sum(operand, axes):
-  return reduce_sum_p.bind(operand, axes=tuple(axes), input_shape=operand.shape)
+  return reduce_sum_p.bind(operand, axes=tuple(axes),
+                           input_shape=onp.shape(operand))
```

Comment:
It's surprising to me to call onp.shape() here. (I'm unclear what onp.shape will do when passed one of our arrays.)

Perhaps define a _shape() function that delegates to onp.shape together with a comment explaining why it works and is needed?

Reply:
Actually we call onp.shape on DeviceArrays and Tracers all the time; if that didn’t do the right thing, we wouldn’t be doing a good job of emulating numpy! (All our _shape functions are just rebindings of onp.shape anyway, to make the name shorter.) All onp.shape does to arraylikes is read their “shape” attr, but it also handles scalars. 

If this were the first appearance of onp.shape used this way it might warrant a comment.

It’s needed because scalars can be passed in here, and scalars can be passed in here because per #121 jax.numpy.array(scalar) returns a scalar (rather than an arraylike with a shape attribute). That is, we don’t have a convenient way to promote scalars to 0D arrays until we fix #121.

Reply:
I double-checked that we use `onp.shape` in other places in lax.py.